### PR TITLE
Added WASI Examples for Rust

### LINKED ISSUE #2: Create some Wasi Demos
So made this an issue for now, due to the new [Runtime in AssemblyScript](https://docs.assemblyscript.org/details/runtime), which changes a lot of things concerning pointers and all of that. And the docs examples seem to not use the new runtime.

Thus, will put this on the backburner for now :smile:

### PR DESCRIPTION
**Pull request by torch2424**
closes #2 

This adds a WASI introduction, as well as a Hello World for Rust, that is similar to the [wasmtime tutorial](https://github.com/bytecodealliance/wasmtime/blob/master/docs/WASI-tutorial.md), but shortened to fit the "by example" theme.

This in general lays the groundwork for WASI on wasm by example :smile: :+1: 

AssemblyScript will come in another PR, as some quality of life improvements will be coming in the next AS release  for WASI: https://github.com/AssemblyScript/assemblyscript/pull/1159

cc @sunfish code, if you got the time, would be excited to hear your thoughts!

Screenshot from 2020-04-24 17-05-22
Screenshot from 2020-04-24 17-05-14
Screenshot from 2020-04-24 17-05-04
Screenshot from 2020-04-24 17-04-54

### CODE REVIEW

File: examples/wasi-hello-world/demo/rust/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// Import rust's io and filesystem module 
+use std::io::prelude::*;
+use std::fs;
+
+// Entry point to our WASI applications
+fn main() {
+    // Print out hello world!
+    // This will handle writing to /dev/stdout for us using the WASI APIs (e.g fd_write)
```
Comment (pchickey) [inline examples/wasi-hello-world/demo/rust/src/main.rs:8]: just `stdout`, its not in /dev/

File: examples/wasi-hello-world/demo/rust/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    println!("Hello world!");
+
+    // Create a file
+    // We are creating a `helloworld.txt` file in the `/helloworld` directory
+    // We will use the wasmtime `--mapdir` flag. 
+    // This will map the `/helloworld` directory on the guest, to  the current directory (`.`) on the host
+    // For example: `--mapdir /helloworld::.`,
+    // This will error if you don't use the `--mapdir` flag correctly.
+    let mut file = fs::File::create("/helloworld/helloworld.txt").unwrap();
```
Comment (pchickey) [inline examples/wasi-hello-world/demo/rust/src/main.rs:17]: I would rephrase this comment as something like:
This code requires the Wasi host to provide a `/helloworld` directory on the guest. If the `/helloworld` directory is not available, the `unwrap()` will cause this program to panic.
For example, in Wasmtime, if you want to map the current directory to `/helloworld`, invoke the runtime with `--mapdir /helloworld::.`.

File: examples/wasi-hello-world/wasi-hello-world.rust.en-us.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
+# WASI Hello World
+
+## Overview
+
+In this example, We will be writing "Hello world!" to both the console (`/dev/stdout`), and a newly created file `./hello-world.txt`. We highly reccomended that you have read the [WASI Introduction](/example-redirect?exampleName=wasi-introduction) before procedding with this example. You should install [wasmtime](https://wasmtime.dev/) as that is the WebAssembly runtime we will be using as our host. You should also ensure that [rustup and rust](https://www.rust-lang.org/tools/install) is installed.
```
Comment (pchickey) [inline examples/wasi-hello-world/wasi-hello-world.rust.en-us.md:5]: `stdout` is just the name of a POSIX io stream (see `man 3 stdout`), the /dev/stdout file is a special Linux thing.

Comment (pchickey) [reply examples/wasi-hello-world/wasi-hello-world.rust.en-us.md:5]: "you should install Rust using the [`rustup`](url here) tool"

File: examples/wasi-hello-world/wasi-hello-world.rust.en-us.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+This example will be similar to [WASI Tutorial on the wasmtime repo](https://github.com/bytecodealliance/wasmtime/blob/master/docs/WASI-tutorial.md).
+
+---
+
+## Implementation
+
+First, we will generate a new binary program with cargo:
```
Comment (pchickey) [inline examples/wasi-hello-world/wasi-hello-world.rust.en-us.md:13]: rather than `binary program` id call it an `executable`

File: examples/wasi-hello-world/wasi-hello-world.rust.en-us.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  // We are creating a `helloworld.txt` file in the `/helloworld` directory
+  // We will use the wasmtime `--mapdir` flag.
+  // This will map the `/helloworld` directory on the guest, to  the current directory (`.`) on the host
+  // For example: `--mapdir /helloworld::.`,
+  // This will error if you don't use the `--mapdir` flag correctly.
+  let mut file = fs::File::create("/helloworld/helloworld.txt").unwrap();
+
+  // Write the text to the file we created
+  write!(file, "Hello world!\n").unwrap();
+}
```
Comment (pchickey) [inline examples/wasi-hello-world/wasi-hello-world.rust.en-us.md:44]: same comments as above on the source file.

File: examples/wasi-hello-world/wasi-hello-world.rust.en-us.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+```
+
+As mentioned in the code comments, even though the [WASI APIs](https://github.com/WebAssembly/WASI/blob/master/phases/snapshot/docs.md) are not being used directly, when we compile our program to WASI, the rust APIs will be using these WASI APIs under the hood for us.
+
+Next, we will want to add WASI as a target that we can compile to. And then compile our program to WASI. To do this we will run:
```
Comment (pchickey) [inline examples/wasi-hello-world/wasi-hello-world.rust.en-us.md:49]: Next, we will ask the `rustup` tool to install support for `WASI`.

File: examples/wasi-introduction/wasi-introduction.all.en-us.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  - When running WebAssembly outside of the server, unlike the browser, you often do not need a full browser engine with support for HTML, CSS, JavaScript, etc... So there are a few popular standalone WebAssembly runtimes / interpreters that you can use as a host, such as:
+    - [Wasmtime](https://wasmtime.dev/)
+    - [Lucet](https://github.com/bytecodealliance/lucet)
+    - [Wasmer](https://wasmer.io/)
+    - [Wasm3](https://github.com/wasm3/wasm3)
+  - Most WebAssembly runtimes / interpreters can be used as a command line interface, or embedded/linked in a larger application by using it's library API.
+  - Each of these projects has their own strengths, and it really depends on what your use case is to choose the best host for you. And there are many more projects out there that you can choose from! For example, here is a [WASI awesome list](https://github.com/wasmerio/awesome-wasi).
+- **"Guests"**
+  - Guests are the WebAssembly modules that are executed in the host. If you plan to write WebAssembly modules, than you would be writing the guest application that is run inside the host application.
```
Comment (pchickey) [inline examples/wasi-introduction/wasi-introduction.all.en-us.md:22]: executed by the host

File: examples/wasi-introduction/wasi-introduction.all.en-us.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  - The host is able to provide additional functionality to guest, by doing tasks on the guests' behalf. This functionality is offered by passing functions to the importObject ([Similar to how we do this for the browser in the "Importing Javascript Functions Into Webassembly"](/example-redirect?exampleName=importing-javascript-functions-into-webassembly)).
+  - This brings us back to WASI, as WASI is a standardized set of APIs for hosts to do system level actions (such as filystem operations) for the guest WebAssembly module. Therefore, this gives guest WebAssembly modules that a developer could write, access to system resources!
```
Comment (pchickey) [inline examples/wasi-introduction/wasi-introduction.all.en-us.md:24]: Therefore, this allows every developer to write WebAssembly modules that can access system resources.

File: examples/wasi-introduction/wasi-introduction.all.en-us.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+The last thing worth mentioning is that WASI uses a [capability based security model](https://github.com/bytecodealliance/wasmtime/blob/master/docs/WASI-capabilities.md). Meaning, hosts must be explicitly given a capability, before it will allow a guest module to perform an action. For example in [Wasmtime](https://wasmtime.dev/), a guest module cannot modify the filesystem, without using the `--mapdir` or `--dir` flag which explictly says which directory on the host (your runtime / computer filesystem) maps to the guest (the webassembly module's virtual filesystem provided by the host).
```
Comment (pchickey) [inline examples/wasi-introduction/wasi-introduction.all.en-us.md:26]: Meaning, host must explicitly grant a capability to a guest module in order for the guest module to perform an action. For example in Wasmtime, by default, the guest module cannot access any part of the host's filesystem. The user that invokes Wasmtime must pass in the `--mapdir` or `--dir` flag to grant modules the capability to access directories in the host filesystem.

File: examples/wasi-introduction/wasi-introduction.all.en-us.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+At the time of this writing, a lot of WASI is still in proposals and things. Only a few features of what WASI is hoping to acheive is fully implemented and standardized. One of those features is filesystem access! Therefore, let's take a look at modifying the file system in the [WASI hello world example](/example-redirect?exampleName=wasi-hello-world), if there is a hello world currently for your language. If not, feel free to look at your language documentation, to see if they support WASI currently, and submit the [WASI hello world example for your language](https://github.com/torch2424/wasm-by-example).
```
Comment (pchickey) [inline examples/wasi-introduction/wasi-introduction.all.en-us.md:28]: It may be worth mentioning "Other system resources, like networking, are not yet part of the WASI standard, though they will be one day. So, if you're hoping to `bind()` to a socket in your WebAssembly module, WASI hosts don't yet expose those capabilities."

**Comment by torch2424** [comment]
Would be great to add a link to: http://www.jakubkonka.com/2020/04/28/rust-wasi-from-scratch.html 

For people interested in how all the wasi calls are made and things

File: examples/wasi-introduction/wasi-introduction.all.en-us.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
+## WASI Introduction
+
+First, we should answer the question, "What is WASI"? WASI is the [WebAssembly System Interface](https://wasi.dev/), "a modular system interface for WebAssembly". To help explain this, let's break it down to:
+
+- **What can WASI do for my WebAssembly Applications?**
+  - WASI offers a [standardized set of APIs](https://github.com/WebAssembly/WASI/blob/master/phases/snapshot/docs.md) for WebAssembly modules to access system resource such as a FileSystem, Networking, Time, Random, etc... .
```
Comment (sunfishcode) [inline examples/wasi-introduction/wasi-introduction.all.en-us.md:6]: My suggestion here would be to mention running wasm "outside the browser" somewhere earlier here, to help new users understand the context here.

File: examples/wasi-introduction/wasi-introduction.all.en-us.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+- **How do WebAssembly modules use WASI?**
+  - A WebAssembly module will use their importObject ([Similar to how we do this for the browser in the "Importing Javascript Functions Into Webassembly"](/example-redirect?exampleName=importing-javascript-functions-into-webassembly)), to be able to access the standardized WASI API bindings (meaning the names of these importObject functions map directly to the agreed-upon WASI implementation).
```
Comment (sunfishcode) [inline examples/wasi-introduction/wasi-introduction.all.en-us.md:8]: `importObject` is a term specific to the JS API, which help users coming from JS, but may also be confusing because in many WASI use cases, there is no JS API. The comparison to the JS tutorial is good, but you may want to clarify that `importObject` is the JS way of doing this. Core wasm itself just has `import`s.

**Comment by torch2424** [comment]
IN general, for the WASI introduction my be worth mentioning:

Without if a user goes straight to the WASI example, without the context of the other examples, they could be a bit lost. Therefore:

* Mention how WASI could be used from a general application perspective (not as someone who writes wasm, which is the current assumption since we assume you are coming from a browser background)
* Mention it's outside of the browser (as already suggested)
* Change the wording to sound more appealing to non web / webassembly developers. Such as the question: "What can WASI do for my WebAssembly Applications?", and try thinking of the question as: "What can WASI do for my Games?" As if, you were a game developer or something.

**Comment by torch2424** [comment]
Made requested chages by @pchickey and @sunfishcode ! Hopefully it looks good, if not, we can always edit it later :smile: :+1: 

Thanks again for the look over, I super appreciate it :)
